### DISCUSSION TITLE
Convert links to footnote with docx output

### DISCUSSION THREAD

**Discussion by ferenci-tamas** [jgm/pandoc#9415] [Q&A] [answered]
Is it possible to achieve the effect of `links-as-notes` but with `docx` output? (I.e., that links are printed as footnotes.) Unfortunately, `links-as-notes` seems to work only for `pdf` output currently.

**Comment by cagix** [comment]
You could probably replace every link with a (link wrapped in a) footnote using a simple Lua filter - both `Note` and `Link` are inline elements (cf. https://pandoc.org/lua-filters.html#type-link).

**Reply by ferenci-tamas** [reply]
Thanks @cagix for the idea! I originally thought that it'll be much more complicated, but this seems to work:

```
return {
  {
    Link = function (link)
        temp = link.content
        temp[#temp + 1] = pandoc.Note {link.target}
        return temp
    end,
  }
}
```

Nevertheless, I'd appreciate if you could check it (is it indeed what you had in your mind?), because I have no experience with Lua.

**Reply by cagix** [reply]
That should do the trick 👍 Maybe @tarleb can weigh in and give some feedback on the Lua coding style?

However, your script would preserve the link text, so it would turn 

```markdown
Lorem ipsum [Pandoc is great](https://pandoc.org/lua-filters.html#examples) foo bar 42.
```

into

```markdown
Lorem ipsum Pandoc is great[^1] foo bar 42.

[^1]: https://pandoc.org/lua-filters.html#examples
```

Is that what you want? I would have expected something like this: 

```markdown
Lorem ipsum[^1] foo bar 42.

[^1]: [Pandoc is great](https://pandoc.org/lua-filters.html#examples)
```

**Reply by ferenci-tamas** [reply]
Thanks for the observation! You're right, but yes, that's what I intended. A realistic example:

`More information can be found in [Pandoc's manual](https://pandoc.org/lua-filters.html#examples) about the filters.`

If this is converted to `docx` with the intention of later being printed to paper then the "Pandoc's manual" part needs the remain in the text so that the sentence is meaningful.

**Reply by cagix** [reply, marked as answer]
@tamas-ferenci You could probably reduce your script to 

```lua
function Link(link)
    link.content:insert(pandoc.Note(link.target))
    return link.content
end
```

**Reply by ferenci-tamas** [reply]
Thanks! That also seems to do the trick (and looks much more elegant!).

Although, surprisingly (to me at least -- again with no background in Lua),

```
function Link(link)
    return link.content:insert(pandoc.Note(link.target))
end
```

does not work.

**Reply by cagix** [reply]
the value of the insert operation is presumably not the new list ...

**Reply by ferenci-tamas** [reply]
Ah, of course! I should have realized it even without Lua knowledge... Thanks, it is clear now.

**Reply by ferenci-tamas** [reply]
I accepted your solution then as an answer!

**Reply by tarleb** [reply]
Sorry for missed the ping, I just found this in my long list of notifications. Since a good solution has already been found, the only thing I can possibly add is some "code golf" to make it shorter. For that, I'd offer this:

``` lua
function Link (link)
    return link.content .. {pandoc.Note{link.target}}
end
```

where `..` is the Lua concatenation operator.

**Comment by aubertc** [comment]
Thanks a lot for the interesting discussion and lua filter. Two notes:
- if the link is itself in a footnote, this will produce invalid docx document,
- links are not clickable.
If I find the time, I'll try to improve and share here a macro that address those issues.

**Comment by aubertc** [comment]
Related? 
https://github.com/cskeeters/pandoc-link-footnote?tab=readme-ov-file

**Comment by aubertc** [comment]
I posted a possible improvement at https://github.com/jgm/pandoc/discussions/11521#discussioncomment-16132866 :

```lua
-- Cf.
-- https://garrit.xyz/posts/2024-04-04-pandoc-convert-links-to-footnotes-the-easy-way
-- https://github.com/jgm/pandoc/discussions/9415
-- https://github.com/jgm/pandoc/discussions/11521

function annotate_links(el)
    return el:walk {
        Link = function (el)
        el.classes:insert("link-in-footnote")
        return el
    end
    }
end

function links_as_footnotes(el)
    if el.classes:includes("link-in-footnote") then
        el.classes:remove(el.classes:find("link-in-footnotes"))
        return el
    else
        el.content:insert(pandoc.Note(pandoc.Para(pandoc.Link(el.target, el.target))))
        return el.content
    end
end

return {
    {Note = annotate_links},
    {Link = links_as_footnotes}
}
```
